pkg/vault/contracts/Vault.sol

```diff
@@ -1185,43 +1187,54 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
-            // Expected amount of wrapped minted by the wrapper protocol.
-            uint256 vaultWrappedDeltaHint;
+            // Expected amount of underlying deposited into the wrapper protocol if buffer is balanced.
+            uint256 vaultUnderlyingDeltaHint = amountInUnderlying;
+            // Expected amount of wrapped minted by the wrapper protocol if buffer is balanced.
+            uint256 vaultWrappedDeltaHint = amountOutWrapped;
 
             if (kind == SwapKind.EXACT_IN) {
                 // The amount of underlying tokens to deposit is the necessary amount to fulfill the trade
                 // (amountInUnderlying), plus the amount needed to leave the buffer rebalanced 50/50 at the end
-                // (bufferUnderlyingSurplus).
-                vaultUnderlyingDeltaHint = amountInUnderlying + bufferUnderlyingSurplus;
+                // (bufferUnderlyingImbalance). `bufferUnderlyingImbalance` may be positive if buffer has an excess
+                // of underlying, or negative if the buffer has an excess of wrapped tokens. `vaultUnderlyingDeltaHint`
+                // will always be a positive number, because if `abs(bufferUnderlyingImbalance) > amountInUnderlying`
+                // and `bufferUnderlyingImbalance < 0`, it means the buffer has enough liquidity to fulfill the trade
+                // (i.e. `bufferBalances.getBalanceDerived() >= amountOutWrapped` is true).
+                vaultUnderlyingDeltaHint = (amountInUnderlying.toInt256() + bufferUnderlyingImbalance).toUint256();
                 underlyingToken.forceApprove(address(wrappedToken), vaultUnderlyingDeltaHint);
                 // EXACT_IN requires the exact amount of underlying tokens to be deposited, so we call deposit.
                 vaultWrappedDeltaHint = wrappedToken.deposit(vaultUnderlyingDeltaHint, address(this));
             } else {
-                if (bufferUnderlyingSurplus > 0) {
-                    // A wrap operation withdraws underlying tokens from the buffer and mints wrapped tokens in return.
-                    // For the buffer to rebalance during a wrap operation, it must have an excess (surplus) of
-                    // underlying tokens. Since this is an EXACT_OUT wrap (mint), we need to calculate the amount of
-                    // wrapped tokens the buffer should receive to rebalance, based on the underlying token surplus.
-                    // The deposit function is used here, as it performs the inverse of a mint operation.
-                    bufferWrappedSurplus = wrappedToken.previewDeposit(bufferUnderlyingSurplus);
+                // We need to calculate both `vaultUnderlyingDeltaHint` and `vaultWrappedDeltaHint`. The underlying
+                // delta will be used to approve transfer from underlying token to the wrapper protocol, while the
+                // wrapped delta will be used to wrap an EXACT_OUT amount of wrapped tokens (mint).
+                if (vaultWrappedDeltaHint != 0) {
```

Comment:
This was a typo (and a very ugly one). I should be checking for the imbalance, not for the delta hint. The tests because it doesn't make a difference in terms of logic, but makes in terms of gas.